### PR TITLE
Translate llvm.fmuladd to fma SPIRV intrinsic

### PR DESCRIPTION
**Pull request by fadeeval**
Translate llvm.fmuladd to fma SPIRV intrinsic

Signed-off-by: Aleksander Fadeev <aleksander.fadeev@intel.com>

### CODE REVIEW

File: lib/SPIRV/SPIRVWriter.cpp [hunk 1]
```diff
@@ -2215,12 +2215,14 @@ SPIRVValue *LLVMToSPIRV::transIntrinsicInst(IntrinsicInst *II,
...
-                          transValue(II->getArgOperand(1), BB), BB);
-    return BM->addBinaryInst(OpFAdd, Ty, Mul,
-                             transValue(II->getArgOperand(2), BB), BB);
+    if (!checkTypeForSPIRVExtendedInstLowering(II, BM))
+      break;
+    std::vector<SPIRVValue *> Ops{transValue(II->getArgOperand(0), BB),
+                                  transValue(II->getArgOperand(1), BB),
+                                  transValue(II->getArgOperand(1), BB)};
+    return BM->addExtInst(transType(II->getType()),
+                          BM->getExtInstSetId(SPIRVEIS_OpenCL), OpenCLLIB::Fma,
```
Comment (bader) [inline lib/SPIRV/SPIRVWriter.cpp:2224]: `mad` seems to be the right instruction to use, instead of `fma`.

File: lib/SPIRV/SPIRVWriter.cpp [hunk 2]
```diff
@@ -2231,12 +2231,14 @@ SPIRVValue *LLVMToSPIRV::transIntrinsicInst(IntrinsicInst *II,
...
-    SPIRVValue *Mul =
-        BM->addBinaryInst(OpFMul, Ty, transValue(II->getArgOperand(0), BB),
-                          transValue(II->getArgOperand(1), BB), BB);
-    return BM->addBinaryInst(OpFAdd, Ty, Mul,
-                             transValue(II->getArgOperand(2), BB), BB);
+    if (!checkTypeForSPIRVExtendedInstLowering(II, BM))
+      break;
+    std::vector<SPIRVValue *> Ops{transValue(II->getArgOperand(0), BB),
+                                  transValue(II->getArgOperand(1), BB),
+                                  transValue(II->getArgOperand(1), BB)};
```
Comment (AlexeySotkin) [inline lib/SPIRV/SPIRVWriter.cpp:2238]:
```suggestion
                                  transValue(II->getArgOperand(2), BB)};
```

File: lib/SPIRV/SPIRVUtil.cpp [hunk 3]
```diff
@@ -1534,8 +1534,10 @@ bool checkTypeForSPIRVExtendedInstLowering(IntrinsicInst *II, SPIRVModule *BM) {
   case Intrinsic::maxnum:
   case Intrinsic::nearbyint: {
     Type *Ty = II->getType();
-    if (II->getArgOperand(0)->getType() != Ty)
-      return false;
+    for (int Index = 0; Index < II->getNumArgOperands(); ++Index) {
+      if (II->getArgOperand(Index)->getType() != Ty)
+        return false;
+    }
```
Comment (AlexeySotkin) [inline lib/SPIRV/SPIRVUtil.cpp:1540]: Probably this check is redundant, I'm not sure it is possible to create these intrinsics with inconsistent argument types.

**Comment by asavonic** [comment]
LGTM.

**Comment by AlexeySachkov** [comment]
I will proceed with #824 instead

**Pull request closed without merging**